hpx/lcos/local/latch.hpp

```diff
@@ -0,0 +1,161 @@
+    /// is a singleuse object; once the operation has been completed, the latch
+    /// cannot be reused.
+    ///
+    /// A latch maintains an internal counter_ that is initialized when the
+    /// latch is created. Threads may block at a synchronization point waiting
+    /// for counter_ to be decremented to 0. When counter_ reaches 0, all such
+    /// blocked threads are released.
+    ///
+    /// Calls to countdown_and_wait() , count_down() , wait() , and is_ready()
+    /// behave as atomic operations.
+    ///
+    /// \note   A \a local::latch is not a LCO in the sense that it has no
+    ///         global id and it can't be triggered using the action (parcel)
+    ///         mechanism. Use lcos::latch instead if this is required.
+    ///         It is just a low level synchronization primitive allowing to
+    ///         synchronize a given number of \a threads.
+    class latch
+    {
+#if defined(BOOST_NO_CXX11_DELETED_FUNCTIONS)
+    private:
+        latch (latch const&);
+        latch (latch&&);
+
+        latch& operator= (latch const&);
+        latch& operator= (latch&&);
+#else
+    public:
+        latch (latch const&) = delete;
+        latch (latch&&) = delete;
+
+        latch& operator= (latch const&) = delete;
+        latch& operator= (latch&&) = delete;
+#endif
+
+    private:
+        typedef lcos::local::spinlock mutex_type;
+
+    public:
+        /// Initialize the latch
+        ///
+        /// Requires: count >= 0.
+        /// Synchronization: None
+        /// Postconditions: counter_ == count.
+        ///
+        explicit latch(std::ptrdiff_t count)
+          : counter_(count)
+        {}
+
+        /// Requires: No threads are blocked at the synchronization point.
+        ///
+        /// \note May be called even if some threads have not yet returned
+        ///       from wait() or count_down_and_wait(), provided that counter_
+        ///       is 0.
+        /// \note The destructor might not return until all threads have exited
+        ///       wait() or count_down_and_wait().
+        /// \note It is the caller's responsibility to ensure that no other
+        ///       thread enters wait() after one thread has called the
+        ///       destructor. This may require additional coordination.
+        ~latch ()
+        {
+            mutex_type::scoped_lock l(mtx_);
+            HPX_ASSERT(counter_ == 0);
+        }
+
+        /// Decrements counter_ by 1 . Blocks at the synchronization point
+        /// until counter_ reaches 0.
+        ///
+        /// Requires: counter_ > 0.
+        ///
+        /// Synchronization: Synchronizes with all calls that block on this
+        /// latch and with all is_ready calls on this latch that return true.
+        ///
+        /// \throws Nothing.
+        ///
+        void count_down_and_wait()
+        {
+            mutex_type::scoped_lock l(mtx_);
+            HPX_ASSERT(counter_ > 0);
```

Comment:
The "Throws Nothing" clause does not conflict with throwing an exception on precondition violation, as that is undefined behavior. Note however that it is not `noexcept`, because somehow `noexcept` beats undefined behavior.